### ISSUE TITLE
Remove-ServiceFabricApplication doesn't work

### ISSUE THREAD

**Issue by jackbond** [Microsoft/service-fabric#279]
When I issue the command 

Remove-ServiceFabricApplication 'myapp' -Force -ForceRemove

the app should be gone in under 15 seconds. No excuses. To have it hang indefinitely is just incomprehensible. And just to be clear, the app I'm trying this on has 38 services, and the cluster is not pegged.

**Comment by GitTorre** [comment]
Hello, Thanks for filing an issue.

Can you elaborate a bit on what you mean by indefinitely. So, Removing the application *never* completes? That sounds like an issue with user code. Without knowing what your 38 services are doing, it's quite hard to diagnose potential blocking issue(s) during shutdown. Things just going poof when you call remove can be hard to guarantee if user code is preventing the service tear downs from making progress.

**Comment by jackbond** [comment]
Let's assume that the services have gone haywire and can't be gracefully removed. Wouldn't that be exactly the situation you'd want to use -Force? As in, "I give up, forget graceful, I just need it gone, kill this thing."

**Comment by GitTorre** [comment]
When we issue a command to take down a service, it is a Ctrl-C, not a kill. So, the user code has a chance to die gracefully, with dignity, to clean up, get ready... -Force doesn't mean anything to us in terms of the command you are using. -Force seems to be a default flag that you can pass along on any PS command. Either way, we don't take it mean to just do something potentially dangerous to user data...

**Comment by jackbond** [comment]
No comment? Is it possible to request that if -Force is supplied than Ctrl-C isn't used? -Force is a pattern in Powershell for a reason, and NOT implementing can very much be considered a bug.

**Comment by jackbond** [comment]
Fascinating...

```
namespace System.Fabric.Description
{
    public sealed class DeleteServiceDescription
    {
        public DeleteServiceDescription(Uri serviceName);

        public Uri ServiceName { get; }
        public bool ForceDelete { get; set; }
    }
}
```

So the DeleteServiceDescription supports kill, which could "do something potentially dangerous to user data", but the Powershell cmdlet for doing THE EXACT SAME OPERATION, and with the EXACT SAME ARGUMENT NAME, doesn't actually set ForceDelete to true?

**Comment by oanapl** [comment]
@jackbond , let me clarify: [Powershell cmdlet](https://docs.microsoft.com/powershell/module/servicefabric/remove-servicefabricservice?view=azureservicefabricps) has 2 parameters, Force (which is the common parameter for all commands) and ForceRemove, which translates into calling delete with the description class you mentioned above with ForceDelete true. So it's all expected. From the documentation:

```powershell
-ForceRemove
Indicates that this command forces removal of the Service Fabric service. This should be used only if the removal of the service is timing out. For example: If Remove-ServiceFabricService is timing out and the replicas of the service are still present then, this means that the service is stuck in ChangeRole/Close and the shutdown sequence cannot complete. By using ForceRemove, we skip the service shutdown sequence and the service is removed ungracefully. For stateful service, its persistent state would not be cleaned. To check if replicas are present, use Get-ServiceFabricReplica.
```

Your initial point was that even with force remove specified, the removal is not done if, for example, the service code has issues and doesn't respond. 
Force remove should always work, though not in under 15 seconds (we don't have a time guarantee for the deletion as of now; however, the deletion should happen in minutes). One situation when it doesn't work is for example, if the message doesn't reach the cluster (user should retry in such case). But other than that, it should go through. We are treating cases when force delete doesn't succeed as bugs, and it's possible you identified a scenario that we didn't cover. We are looking into one such improvement in the next release, but that may not be your case.

To summarize: you should never have to delete the resource group for a misbehaving app. If you see this again, please send us the cluster information (cluster resource ID and region) and information about the app and we can investigate some more. Or open a support ticket and we can collect the data we need.

Thank you for your feedback and support!
